File: services/backend/app/api/assessments.py
from fastapi import APIRouter, Depends
from sqlalchemy.orm import Session

from app.api.deps import get_current_user
from app.core.database import get_db
from app.models.assessment import AssessmentAnchor
from app.models.user import User
from app.schemas.assessment import AssessmentAnchorCreate, AssessmentAnchorRead
from app.schemas.kbat_result import KBatResultResponse
from app.services.kbat_result import gather_kbat_result

router = APIRouter()


@router.post("/assessments/anchor", response_model=AssessmentAnchorRead, status_code=201)
def create_assessment_anchor(
    payload: AssessmentAnchorCreate,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
) -> AssessmentAnchor:
    anchor = AssessmentAnchor(
        user_id=current_user.id,
        assessment_type=payload.assessment_type,
        target_group=payload.target_group,
        completed_at=payload.completed_at,
        dimensions=payload.dimensions,
        interpretation_scope=payload.interpretation_scope,
        source=payload.source,
        supersedes_id=payload.supersedes_id,
    )
    db.add(anchor)
    db.commit()
    db.refresh(anchor)
    return anchor


@router.get("/assessments/anchor", response_model=AssessmentAnchorRead)
def read_latest_assessment_anchor(
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
) -> AssessmentAnchor:
    latest = (
        db.query(AssessmentAnchor)
        .filter(AssessmentAnchor.user_id == current_user.id)
        .order_by(AssessmentAnchor.created_at.desc())
        .first()
    )
    if latest is None:
        from fastapi import HTTPException
        raise HTTPException(status_code=404, detail="저장된 검사 응답이 없습니다")
    return latest


@router.get("/assessments/kbat-result", response_model=KBatResultResponse)
def read_kbat_result(
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
) -> KBatResultResponse:
    """K-BAT 자가진단(설문) + 누적 일상기록을 합친 최종 결과.

    설문을 아직 안 했거나(NOT_TAKEN) 일상기록이 7일 미만(INSUFFICIENT_
    RECORDS)이면 항상 200으로 그 상태를 알려준다 -- "결과가 아직 없음"은
    오류가 아니라 정상 상태이므로 404로 표현하지 않는다(#137/#138과 같은
    원칙).
    """
    snapshot = gather_kbat_result(db, user_id=current_user.id)
    return KBatResultResponse(
        state=snapshot.state,
        recorded_days=snapshot.recorded_days,
        minimum_required_days=snapshot.minimum_required_days,
        survey_completed_at=snapshot.survey_completed_at,
        result=snapshot.result,
    )

File: services/frontend/User dashboard/src/app/api/assessments.ts
import { apiFetch } from "./client";
import { getCurrentUser } from "./users";

// app/models/assessment.py의 AssessmentType / InterpretationScope
export type AssessmentType = "custom_initial_state_survey" | "k_bat";
export type InterpretationScope = "fixed_reference_only";

// app/models/user.py의 UserType — 검사 결과를 어떤 집단 기준으로 볼지 정한다
export type UserType = "university_student" | "job_seeker" | "early_career_worker";

// app/schemas/assessment.py의 DimensionKey와 1:1.
// 앞 4개는 온보딩 초기 상태 설문용, 뒤 3개는 K-BAT 하위영역용이고
// exhaustion(탈진)은 두 검사가 공유한다.
export type DimensionKey =
  | "exhaustion"
  | "academic_burden"
  | "occupational_burden"
  | "recovery_difficulty"
  | "mental_distance"
  | "cognitive_control"
  | "emotional_control";

/** 차원별 0(부담 없음)~1(부담 큼) 점수. 응답하지 않은 차원은 null. */
export type DimensionScores = Partial<Record<DimensionKey, number | null>>;

export interface AssessmentAnchorCreate {
  assessment_type: AssessmentType;
  target_group: UserType;
  completed_at: string; // ISO8601, 타임존 필수 -- new Date().toISOString()
  dimensions: DimensionScores;
  source: string;
  supersedes_id?: string | null;
}

export interface AssessmentAnchorRead extends AssessmentAnchorCreate {
  id: string;
  user_id: string;
  interpretation_scope: InterpretationScope;
  created_at: string;
}

const USER_TYPES: readonly UserType[] = ["university_student", "job_seeker", "early_career_worker"];

/** 검사 저장에 필수인 target_group을 알아낸다.
 *  user_type이 아직 비어있거나(온보딩 중 계정) 호출이 실패하면 null --
 *  자가진단 화면이 저장 때문에 막히면 안 되므로 호출부가 저장을 건너뛴다. */
export async function getTargetGroup(): Promise<UserType | null> {
  try {
    const user = await getCurrentUser();
    return USER_TYPES.includes(user.user_type as UserType) ? (user.user_type as UserType) : null;
  } catch {
    return null;
  }
}

export async function createAssessmentAnchor(
  payload: AssessmentAnchorCreate
): Promise<AssessmentAnchorRead> {
  return apiFetch<AssessmentAnchorRead>("/assessments/anchor", {
    method: "POST",
    body: JSON.stringify(payload),
  });
}
